File: 51_N-Queens.java
class Solution {

    public List<List<String>> solveNQueens(int n) {

        /* Sanity check */
        if(n == 0)
            return new ArrayList<List<String>>();

        ArrayList<List<String>> result = new ArrayList<>();
        int[] pos = new int[n];

        dfs(result, pos, 0);

        return result;
    }

    /* dfs */
    private void dfs(List<List<String>> result, int[] pos, int index) {
        /* One solution found */
        if(index == pos.length) {
            result.add(getResultString(pos));
            return;
        }

        /* Keep searching */
        for(int i = 0; i < pos.length; ++i) {
            pos[index] = i;

            if(isConflict(pos, index) == false)
                dfs(result, pos, index + 1);
        }

    }

    /* getResultString - Get the string form of a valid solution */
    private List<String> getResultString(int[] pos) {
        ArrayList<String> result = new ArrayList<>();
        StringBuilder sb = new StringBuilder();
        for(int i = 0; i < pos.length; ++i)
            sb.append(".");

        for(int i = 0; i < pos.length; ++i) {
            sb.setCharAt(pos[i], 'Q');
            result.add(sb.toString());
            sb.setCharAt(pos[i], '.');
        }

        return result;
    }

    /* isConflict - Returns whether column i conflicts with previous columns. */
    private boolean isConflict(int[] pos, int index) {
        for(int i = 0; i < index; ++i) {
            if(pos[index] == pos[i] || (index - i) == Math.abs(pos[index] - pos[i]))
                return true;
        }
        return false;
    }

}

File: 51_N-Queens.cpp
class Solution {
 private:
  bool conflict(const vector<int>& position, int i) {
    for (int j = 0; j < i; ++j) {
      if (position[i] == position[j] ||
          (i - j) == abs(position[i] - position[j]))
        return 1;
    }
    return 0;
  }

  void print(const vector<int>& position, vector<vector<string>>& result) {
    int n = position.size();
    vector<string> temp(n, string(n, '.'));
    for (int i = 0; i < n; ++i) {
      temp[position[i]][i] = 'Q';
    }
    result.push_back(temp);
  }

  void dfs(vector<int>& position, vector<vector<string>>& result, int i) {
    // cout << "dfs executed, depth: " << i << endl;
    int n = position.size();
    if (i == n) {  // solution found
      print(position, result);
      return;
    }
    for (int j = 0; j < n; ++j) {
      position[i] = j;
      if (!conflict(position, i)) dfs(position, result, i + 1);
    }
  }

 public:
  vector<vector<string>> solveNQueens(int n) {
    vector<int> position(n);
    vector<vector<string>> result;
    dfs(position, result, 0);
    return result;
  }
};